src/coordinator/block/interface.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/models"
+)
+
+const (
+	// errBounds is returned when time requested is outside the block bounds
+	errBounds = "out of bounds, time: %v, bounds: %v"
+)
+
+// Block represents a group of series across a time bound
+type Block interface {
+	Meta() Metadata
+	StepIter() StepIter
+	SeriesIter() SeriesIter
+	SeriesMeta() []SeriesMeta
+	Close() error
+}
+
+// SeriesMeta is metadata data for the series
+type SeriesMeta struct {
+	Tags models.Tags
+}
+
+// Bounds are the time bounds
+type Bounds struct {
+	Start    time.Time
+	End      time.Time
+	StepSize time.Duration
+}
+
+// SeriesIter iterates through a CompressedSeriesIterator horizontally
+type SeriesIter interface {
+	Next() bool
+	Current() Series
+}
+
+// StepIter iterates through a CompressedStepIterator vertically
+type StepIter interface {
+	Next() bool
+	Current() Step
+	Steps() int
+}
+
+// Step can optionally implement iterator interface
+type Step interface {
+	Time() time.Time
+	Values() []float64
+}
+
+// Metadata is metadata for a block
+type Metadata struct {
+	Bounds Bounds
+	Tags   models.Tags // Common tags across different series
+}
+
+// Builder builds a new block
+type Builder interface {
+	AppendValue(index int, value float64) error
+	Build() Block
+	AddCols(num int) error
+}
+
+// Result is the result from a block query
+type Result struct {
+	Blocks []Block
+}
+
```

Comment:
nit `var EmptyResult = Result{}`, then use block.EmptyResult wherever you're returning `block.Result{}`

Reply:
always returning block.Result{} should be cheap enough since it's allocated an empty struct on the stack and then copying it over to the calling function.